api/v1alpha1/containerimagebuild_types.go

```diff
@@ -46,19 +24,68 @@ type Registry struct {
 	SecretNamespace string `json:"secretNamespace"`
 }
 
+func (auth BasicAuthConfig) Validate() error {
+	username := auth.Username != ""
+	password := auth.Password != ""
+	secretName := auth.SecretName != ""
+	secretNamespace := auth.SecretNamespace != ""
+
+	switch {
+	case !username && !password && !secretName && !secretNamespace: // no basic auth provided
+		return nil
+	case (username && !password) || (!username && password): // inline auth validation
+		return errors.New("inline basic auth requires both username and password")
+	case (secretName && !secretNamespace) || (!secretName && secretNamespace): // secret auth validation
+		return errors.New("secret basic auth requires both secret name and namespace")
+	case (username && (secretName || secretNamespace)) ||
+		(password && (secretName || secretNamespace)) ||
+		(secretName && (username || password)) ||
+		(secretNamespace && (username || password)):
+		return errors.New("basic auth cannot be both inline and secret-based")
+	}
```

Comment:
Hmm ... this is rough if we ever need to add / modify fields, can we just XOR IsInline & IsSecret and make a generic error message?

Reply:
The XOR will handle the case where both types of "full credentials" are provided and clean up the last `case` statement, but it will not guard against _partial credentials_. Ultimately, this is fine because a build will fail if credentials are required but not provided in-full, but do we not want to guard against that prior to launching a build (expensive)?

I think ignoring cases like `username != "" && (secretName != "" || secretNamespace != "")` is fine but I wonder if we should keep the partial credentials check. Check out 150d4b6f17f53584ef9f4ee592053d51cbe36951 and let me know what you think (tests included).